sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,505 @@
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>Whether the submission was throttled.</returns>
+        protected async Task SubmitBatchAsync(IList<PublisherAction<T>> batch, CancellationToken cancellationToken)
+        {
+            // If Submit is called before our last retry delay elapsed, we'll
+            // wait before sending the next request
+            await _manualRetries.WaitIfNeededAsync(cancellationToken).ConfigureAwait(false);
+
+            // Send the request
+            bool throttled = await OnSubmitBatchAsync(batch, cancellationToken).ConfigureAwait(false);
+
+            // Update whether or not the request was throttled to update our
+            // retry delay
+            _manualRetries.Update(throttled);
+        }
+
+        /// <summary>
+        /// Send indexing actions to be processed by the service.
+        /// </summary>
+        /// <param name="batch">The batch of actions to submit.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>Whether the submission was throttled.</returns>
+        protected abstract Task<bool> OnSubmitBatchAsync(IList<PublisherAction<T>> batch, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Enqueue an action to be retried.
+        /// </summary>
+        /// <param name="action">The action to be retried.</param>
+        /// <param name="skipIncrement">
+        /// Whether we should skip incrementing the retry attempts because the
+        /// failure wasn't related to the action.
+        /// </param>
+        /// <returns>
+        /// A value indicating whether the action is retriable.
+        /// </returns>
+        protected bool EnqueueRetry(
+            PublisherAction<T> action,
+            bool skipIncrement = false)
+        {
+            bool retriable = skipIncrement || action.RetryAttempts++ < RetryCount;
+            if (retriable)
+            {
+                _retry.Enqueue(action);
+            }
+            return retriable;
+        }
+        #endregion Publishing
+
+        #region Timer
+        /// <summary>
+        /// Start the timer if it's not already running.
+        /// </summary>
+        private void StartTimer()
+        {
+            if (AutoFlushInterval != null)
+            {
+                int intervalInMs = (int)AutoFlushInterval.Value.TotalMilliseconds;
+                if (_timer == null)
+                {
+                    _timer = new Timer(
+                        OnTimerElapsed,
+                        state: null,
+                        dueTime: intervalInMs,
+                        period: Timeout.Infinite);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stop the timer.
+        /// </summary>
+        private void StopTimer()
+        {
+            if (_timer != null)
+            {
+                _timer.Dispose();
```

Comment:
You don't have to dispose the timer to pause it.

`timer.Change(Timeout.Infinite, Timeout.Infinite);`

Reply:
Any idea how to tell if a timer is in that state?  We don't want to restart it from scratch so I think it'll take another bool to track the state.  I'll make that change instead of disposing though.

Reply:
I don't know a way to do that, unfortunately.